File: backend/src/main/java/com/example/kolla/services/SearchService.java
package com.example.kolla.services;

import com.example.kolla.dto.MeetingSearchResult;
import com.example.kolla.dto.SearchMeetingRequest;
import com.example.kolla.dto.TranscriptionSearchResult;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

/**
 * Service interface for search operations.
 * Requirements: 13.1–13.7
 */
public interface SearchService {

    /**
     * Search meetings with dynamic filters.
     * Supports filtering by keyword (title/description), date range, room, department, and creator.
     * Requirements: 13.1–13.3
     *
     * @param request  filter criteria (all fields optional)
     * @param pageable pagination and sort parameters
     * @return paginated list of matching meeting search results
     */
    Page<MeetingSearchResult> searchMeetings(SearchMeetingRequest request, Pageable pageable);

    /**
     * Full-text search in transcription segment text.
     * Requirements: 13.4–13.7
     *
     * @param keyword   required search term (matched case-insensitively against segment text)
     * @param meetingId optional meeting ID to scope the search
     * @param pageable  pagination and sort parameters
     * @return paginated list of matching transcription segment results
     */
    Page<TranscriptionSearchResult> searchTranscriptions(String keyword, Long meetingId,
                                                          Pageable pageable);
}

File: frontend/src/services/searchService.ts
/**
 * searchService.ts — search meetings and transcriptions.
 * Requirements: 13.1–13.7
 */

import api from './api'
import type { ApiResponse, PageResponse } from '../types/api'
import type { Meeting } from '../types/meeting'

export interface MeetingSearchFilters {
  query?: string
  startDate?: string
  endDate?: string
  roomId?: number
  departmentId?: number
  page?: number
  size?: number
}

export interface TranscriptionSearchResult {
  segmentId?: number
  jobId?: string
  meetingId: number
  meetingTitle: string
  speakerName: string
  text: string
  segmentStartTime: string
}

export interface TranscriptionSearchFilters {
  query: string
  page?: number
  size?: number
}

/**
 * Search meetings by date range, room, department, and text query.
 * Requirements: 13.2, 13.5
 */
export async function searchMeetings(
  filters: MeetingSearchFilters,
): Promise<ApiResponse<PageResponse<Meeting>>> {
  const params: Record<string, string | number | undefined> = {}
  if (filters.query?.trim()) params.keyword = filters.query.trim()
  if (filters.startDate) params.startDate = filters.startDate
  if (filters.endDate) params.endDate = filters.endDate
  if (filters.roomId !== undefined) params.roomId = filters.roomId
  if (filters.departmentId !== undefined) params.departmentId = filters.departmentId
  if (filters.page !== undefined) params.page = filters.page
  if (filters.size !== undefined) params.size = filters.size

  const response = await api.get<ApiResponse<PageResponse<Meeting>>>('/search/meetings', { params })
  return response.data
}

/**
 * Full-text search in transcription segments.
 * Requirements: 13.4, 13.6
 */
export async function searchTranscriptions(
  filters: TranscriptionSearchFilters,
): Promise<ApiResponse<PageResponse<TranscriptionSearchResult>>> {
  const params: Record<string, string | number | undefined> = {
    keyword: filters.query.trim(),
  }
  if (filters.page !== undefined) params.page = filters.page
  if (filters.size !== undefined) params.size = filters.size

  const response = await api.get<ApiResponse<PageResponse<TranscriptionSearchResult>>>(
    '/search/transcriptions',
    { params },
  )
  return response.data
}
